common/deadlock/deadlock.go

```diff
@@ -0,0 +1,222 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deadlock
+
+import (
+	"context"
+	"runtime/pprof"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/fx"
+	"google.golang.org/grpc/health"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/internal/goro"
+	"go.temporal.io/server/service/history/shard"
+)
+
+type (
+	params struct {
+		fx.In
+
+		Logger       log.SnTaggedLogger
+		Collection   *dynamicconfig.Collection
+		HealthServer *health.Server
+
+		// root pingables:
+		NamespaceRegistry namespace.Registry
+		ClusterMetadata   cluster.Metadata
+		ShardController   shard.Controller `optional:"true"`
```

Comment:
why is shard controller optional?

Reply:
oh, is it because this is run for all server roles, and shard controller might not be available for other roles.

Reply:
we probably should check matching.taskQueuesLock as well. Can be in follow up PR if you prefer.

Reply:
I was going to do that and looked at the code, and it's so simple I could verify by sight that it can't deadlock (I think). but might as well add it, it doesn't hurt.